src/eigen/preconditioners.jl

```diff
@@ -11,6 +11,19 @@
 import LinearAlgebra: ldiv!
 import LinearAlgebra: mul!
 
+# Returns dot(A[:, i], Diagonal(P.kin), B[:, i])), for all columns of A, B
```

Comment:
Mismatch between comment and implementation.

Also I would actually do this like the `dot` function. I.e. have a two-argument version (where you just call the two-argument `dot`) and a three-argument version, where you call the three-argument `dot`. I think the three-argument `dot` with an `I` (uniform scaling) compiles away, so you can just have that as your implementation and then dispatch `diag_prod(A, B) = diag_prod(A, I, B)`.